0000-classical-stores.md

```diff
@@ -0,0 +1,493 @@
+    def add_capture(self, *vars: expr.Var): ...
+    @overload
+    def add_input(self, name_or_var: str, type: types.Type, /) -> expr.Var: ...
+    @overload
+    def add_input(self, name_or_var: expr.Var, /) -> expr.Var: ...
+
+    def store(self, location: expr.Expr, value: expr.Expr, /): ...
+
+    @overload
+    def get_var[T](self, name: str, *, default=T) -> Union[expr.Var, T]: ...
+    @overload
+    def get_var(self, name: str, *, default=NotGiven) -> expr.Var: ...
+    def has_var(self, var: Union[str | expr.Var]) -> bool: ...
+
+    # All variables, including inputs and captures.
+    def iter_vars(self) -> Iterable[expr.Var]: ...
+
+    # All do what they say on the tin.
+    def iter_declared_vars(self) -> Iterable[expr.Var]: ...
+    def iter_input_vars(self) -> Iterable[expr.Var]: ...
+    def iter_captured_vars(self) -> Iterable[expr.Var]: ...
+```
+
+`QuantumCircuit.add_var` and `.add_input` are for declaring a variable with an owned storage location, the difference being that `add_var` is an internal variable and requires an initialiser, while `add_input` marks this as in input variable to the circuit and cannot have an initialiser.
+The variable remains in scope for the whole body of the circuit.
+Variables are not automatically inherited by nested scopes; they must be explicitly closed over when building an inner scoped `QuantumCircuit` instance using `QuantumCircuit.add_capture`.
+Only `add_input` takes a `Type`, and even then, only if the input is given as a `str` name.
+All other forms read the type from the `expr.Var` or the initialiser `expr.Expr`.
+It is an error if the initialiser expression is not of a type that implicitly casts to the type of the `Var`.
+If a `str` name is given in `add_var`, the type of the resulting expression will be exactly the type of the initialiser, and this should be manually cast to the desired type if it is not the exact required type.
+
+The initialiser of `QuantumCircuit` is expanded to allow setting these as well.
+This is not intended to be how the _first_ circuit a user creates constructs its variables in general; it is meant to make it easy to build more circuits similar to the first circuit by reusing the same `Var` instances that are already held.
+In the initialiser, all the variables must be given as `expr.Var`, since if a string was permitted, a user would then have to do `get_var` to retrieve the `Var` before they could do anything anyway, at which point they can use `add_var` which is just as ergonomic.
+The keyword `declarations` must take some form of mapping from the `expr.Var` values to the initial expressions.
+A user must supply a mapping or iterable that will yield the `Var` nodes in a suitable order for declaration; it will be an error if an initialiser expression relies on a `Var` that has not yet been seen.
+Python's built-in `dict` yields its items in insertion order (up to removals), so this is not expected to be an onerous ask for the API.
```

Comment:
Does python guarantee this for `dict` or is it just an artifact of the current implementation?

Reply:
It used to be a CPython implementation detail, but since 3.7 it's been a part of the language spec.